src/server/engine_shard_set.cc

```diff
@@ -56,6 +318,31 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+void EngineShard::StartDefragTask(util::ProactorBase* pb) {
+  MemoryDefragParams params(mi_resource_.heap(), &db_slice_);
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  defrag_task_ = pb->AddOnIdleTask(
+      [task_params = std::move(params), commited, mem_in_use, count = 0]() mutable {
```

Comment:
Why not make this function a member function? Then you can get rid of `MemoryDefragParams`. it's simpler and more readable.

Reply:
I not sure params required any more, for the lambda we need to state of the current memory usage, so it can call a member function from inside by passing this to the lambda but you still have the lambda state that should not "leak" EngineShard class